src/search3_indexer.erl

```diff
@@ -0,0 +1,221 @@
+-module(search3_indexer).
+
+-export([
+    update/2
+]).
+
+-export([
+    spawn_link/0
+]).
+
+
+-export([
+    init/0
+]).
+
+-include("search3.hrl").
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("fabric/include/fabric2.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-import(couch_query_servers, [
+    get_os_process/1,
+    ret_os_process/1,
+    proc_prompt/2
+]).
+
+spawn_link() ->
+    proc_lib:spawn_link(?MODULE, init, []).
+
+init() ->
+    {ok, Job, JobData} = couch_jobs:accept(?SEARCH_JOB_TYPE, #{}),
+    #{
+        <<"db_name">> := DbName,
+        <<"ddoc_id">> := DDocId,
+        <<"name">> := IndexName,
+        <<"sig">> := JobSig
+    } = JobData,
+    {ok, Db} = fabric2_db:open(DbName, []),
+    {ok, DDoc} = fabric2_db:open_doc(Db, DDocId),
+    {ok, Index} = search3_util:design_doc_to_index(DDoc, IndexName),
+    Index1 = Index#index{dbname = DbName},
+    HexSig = fabric2_util:to_hex(Index1#index.sig),
+
+    if  HexSig == JobSig -> ok; true ->
+        couch_jobs:finish(undefined, Job, JobData#{
+            error => sig_changed,
+            reason => <<"Design document was modified">>
+        }),
+        exit(normal)
+    end,
+    State = #{
+        txdb => undefined,
+        search_seq => undefined,
+        count => 0,
+        limit => config:get_integer("search3", "change_limit", 100),
+        doc_acc => [],
+        last_seq => undefined,
+        job => Job,
+        job_data => JobData,
+        index => Index1,
+        proc => undefined
+    },
+    update(Db, State).
+
+update(#{} = Db, State) ->
+    try
+        Index = maps:get(index, State),
+        Seq = search3_rpc:get_update_seq(Index),
+        Proc = get_os_process(Index#index.def_lang),
+        NewState = State#{
+            search_seq => Seq,
+            last_seq => Seq,
+            proc => Proc
+        },
+        try
+            true = proc_prompt(Proc, [<<"add_fun">>, Index#index.def]),
+            update_int(Db, NewState)
+        after
+            ret_os_process(Proc)
+        end
+    catch error:database_does_not_exist ->
+        #{db_prefix := DbPrefix} = Db,
+        couch_log:error("Db - ~p database does not exist", [DbPrefix]),
+        erlang:error(database_does_not_exist)
+    end.
+
+update_int(#{} = Db, State) ->
+    State3 = fabric2_fdb:transactional(Db, fun(TxDb) ->
+        State1 = maps:put(tx_db, TxDb, State),
```

Comment:
Can use `State1 = State#{tx_db := TxDb}` here. It would also assert that tx_db should exist in the map, so we don't accidentally create a new key you didn't mean to create, say txdb vs tx_db.